pdal/pio.py

```diff
@@ -0,0 +1,108 @@
+import types
+import json
+import subprocess
+from functools import partial
+from collections import defaultdict
+from itertools import chain
+import copy
+import warnings
+
+import pdal
+
+try:
+    PDAL_DRIVERS_JSON = subprocess.run(["pdal", "--drivers", "--showjson"], capture_output=True).stdout
+    PDAL_DRIVERS = json.loads(PDAL_DRIVERS_JSON)
+    _PDAL_VALIDATE = True
+except:
+    PDAL_DRIVERS = []
+    _PDAL_VALIDATE = False
+
+DEFAULT_STAGE_PARAMS = defaultdict(dict)
+DEFAULT_STAGE_PARAMS.update({
+# TODO: add stage specific default configurations
+})
+
+
+class StageSpec(object):
+    def __init__(self, prefix, **kwargs):
+        self.prefix = prefix
+        self.key = ".".join([self.prefix, kwargs.get("type", "")])
+        self.spec = DEFAULT_STAGE_PARAMS[self.key].copy()
+        self.spec.update(kwargs)
+        self.spec["type"] = self.key
+        # NOTE: special case to support reading files without passing an explicit reader
+        if (self.prefix == "readers") and kwargs.get("type") == "auto":
+            del self.spec["type"]
+
+    @property
+    def pipeline(self):
+        output = PipelineSpec()
+        output.add_stage(self)
+        return output
+
+    def __getattr__(self, name):
+        if _PDAL_VALIDATE and (name not in dir(self)):
+            raise AttributeError(f"'{self.prefix}.{name}' is an invalid or unsupported PDAL stage")
+        return partial(self.__class__, self.prefix, type=name)
```

Comment:
This seems a little magic to me, but I'm not a python guy.  Can you explain?  What type does the caller of `__getattr__ `expect?  (Not that this is a comment on this PR).

Reply:
Yeah this is definitely the magic.  When `__getattr__` gets called implicitly by attribute lookup (ie `pio.readers.las` or whatever), this method gets called with `name="las"`.  It then returns a callable in order to support the functional syntax of `pio.readers.las(filename="myfile.las").  The callable that gets returned is a version of the constructor with some of the parameters locked in ("curried function" is maybe the term, can never remember), which allows us to get the open ended api without building an explicit hierarchy.  The prefix attribute is there to work around the dot notation in stage names without building an empty class hierarchy.  There's def other ways, but not immediately certain any are much less magical in Python unfortunately.

Reply:
Thanks for the nice explanation.